### ISSUE TITLE
Control Flow Integrity

### ISSUE THREAD

**Issue by defuse** [zcash/zcash#3810]
Control Flow Integrity (CFI) is an approach to mitigating code-reuse attacks (e.g. ROP). In this ticket, evaluate the CFI options that are available and see if any would be beneficial to `zcashd`. A quick search returns:

- [Microsoft's Control-Flow Guard](https://docs.microsoft.com/en-us/windows/desktop/secbp/control-flow-guard) (used in Edge)
- [LLVM's Control Flow Integrity](https://clang.llvm.org/docs/ControlFlowIntegrity.html) (used in Chrome)

**Comment by daira** [comment]
Microsoft CFG seems inapplicable unless the code is compiled with MSVC, which we don't support, and I think we're unlikely to support.

We do support compiling with clang, at least for Thread/Address Sanitizer builds. (I don't know whether those still require a patch, but if so it wouldn't be difficult to fix.)

**Comment by daira** [comment]
It looks like we have to solve #2513 first in order for the [cfi-mfcall](https://clang.llvm.org/docs/ControlFlowIntegrity.html#member-function-pointer-call-checking) check to work:

> For this scheme to work, all translation units containing the definition of a virtual member function (whether inline or not), other than members of blacklisted types or types with public LTO visibility, must be compiled with ``-flto`` or ``-flto=thin`` enabled and be statically linked into the program.

because libstdc++ will contain such virtual member functions.

(Note that #2513 is currently blocked by #3744.)

**Comment by zebambam** [comment]
We can still enable the other checks though, right?

**Comment by zebambam** [comment]
> If I understand correctly, as long as we still dynamically link to the system libc, this won't prevent [return-to-libc attacks](https://en.wikipedia.org/wiki/Return-to-libc_attack), and similarly for libstdc++ and other dynamically linked system libraries (see #2513).

Right, but ASLR puts standard libraries at random locations, making ret-to-libc harder. 

In general the incompleteness of one mitigation technology shouldn't remove it from our consideration.  All of them are incomplete, we should consider them and their effect as a set, and one that we improve over time.
